### DISCUSSION TITLE
How to get the list of boundary events of an activity.

### DISCUSSION THREAD

**Discussion by mahdiSaeediN** [imixs/open-bpmn#310] [Q&A]
Hi!
Thanks for all the great work in this project.

I'm new to open-bpmn. In a project I need to modify bpmn models and save new instances. For that, I need to traverse through all nodes in a given model. An issue now I encountered is to access the list of boundary events for an activity. I'm able to get the list of activites in a process using _process.getActivities()_, but I don't see a way to access its boundary events. Same is for some other elements related to the node like its data objects. Could you help me with this issue?

Other questions: 
Does the API cover all BPMN elements now?
Is there an API documentation that I can use?

Thanks.

**Comment by rsoika** [comment]
Hi  @mahdiSaeediN , I am not absolute sure but I think the `getActivities` method includes also the boundaryEvents.
You can test the element by its type:


```java
for (Activity activity : process.getActivities()) {
   if (BPMNTypes.BOUNDARY_EVENT.equals(activity.getType())) {
   ....
```

Can you test this? If it works we can add a convenience method `getBoundaryEvents` to the `Process` instance. 

We do not cover all elements now. The BPMNModel provide generic methods to navigate through a model, like `findChildNodesByName` or `createElement`, but for this you need to know the internal BPMN logic in more detail. These methods are public and used internally by the Open-BPMN Meta Model to navigate through the model. In general you always have full access to the model because the Meta Model allows you to read and write in the BPMN file.

**Reply by mahdiSaeediN** [reply]
Hi @rsoika, boundaryEvent is actually an event type. I can find the list of boundary events in a process using below code:
```java
for (Event e : process.getEvents()) {
          if (BPMNTypes.BOUNDARY_EVENT.equals(e.getType())) {
           System.out.println(e.getName());
          }
      }
```
but I'm looking for a method to get boundary events of a given activity (like _activity.getBoundaryEvents()_). Same is for its data objects.

Do you have a list of elements that is covered now? because I need to know about the coverage so to make sure if I can use it for my research. 

You mentioned that the coverage is limited but we have full access to the model. You mean we can access every elemenet of the model but we can't add/remove/modify all elements?

Another question: How does you API compare to the Camunda BPMN API?

Thanks.

**Reply by rsoika** [reply]
> but I'm looking for a method to get boundary events of a given activity (like activity.getBoundaryEvents()). Same is for its data objects.

OK, good point - I have opened a [new Issue](https://github.com/imixs/open-bpmn/issues/311) for this feature. 

> You mentioned that the coverage is limited but we have full access to the model. You mean we can access every elemenet of the model but we can't add/remove/modify all elements?

you have full access to all elements via the XML Dom tree. The main goal of the Open-BPMN meta project is just to provide you with a set of convenience methods to make creation, reading and modifying a model more easy. The methods you see today are the ones used by the Open-BPMN graphical modeller - based on the [GLSP project](https://github.com/eclipse-glsp/). Currently this is more a lack of documentation...

> Another question: How does you API compare to the Camunda BPMN API?

I personal do not think, that Camunda offers a comparable open concept.

**Comment by rsoika** [comment]
@mahdiSaeediN I just added new methods to create and read BoundaryEvents from an Activity. 

Take a look here: https://github.com/imixs/open-bpmn/blob/master/open-bpmn.metamodel/src/test/java/org/openbpmn/metamodel/examples/TestCreateEvents.java

Is this what you expect?
